writers/cap.go

```diff
@@ -0,0 +1,16 @@
+package writers
```

Comment:
Can we drop this abstraction?
I'm not sure it's necessary to wrap an integer this way (also it seems we don't need it to be generic as we always use `int64`).

Reply:
> We can make it `int64` always, but we have a lot of duplication for this, so I'd rather leave this.

I don't think the abstraction removes the duplicate it just wraps it in `Add` instead of `+=`, `GetCurrent` instead of direct access

Reply:
Also `OverflownBy` instead of `>`

Reply:
> Also `OverflownBy` instead of `>`

`max > 0 && curr + delta > max`, but yes.

Still, I feel like it's a good abstraction & I haven't seen any argument why not (we basically get to keep the limit & the curr value coupled together with distinct meaning).

Reply:
> I haven't seen any argument why not

You have to learn the abstraction to understand how the code works

> we basically get to keep the limit & the curr value coupled together with distinct meaning

We already have them together in the function scope?

Reply:
Also if you want to update the code you need to modify the abstraction so it breaks fast (e.g. `OverflownBy` to keep the original order of the operations)

Reply:
I've added the `OverflownBy` now & removed `ReachedLimit`

Reply:
> I've added the `OverflownBy` now & removed `ReachedLimit`

Sorry for not being clear, that was my point. To fix the bug you had to break the abstraction on the first change. Which probably means it's not ready to be abstracted yet

Reply:
I'm not ready to die on the hill for this impl, so sure, I'll remove it.